examples/common/hf_fsdp.py

```diff
@@ -0,0 +1,220 @@
+    }
+
+    for mod_name, module in modules.items():
+        if module is None:
+            raise ValueError(
+                f'Unable to FSDP-wrap this model! `{mod_name}` does not ' +
+                'follow common layer/weight naming conventions.')
+    block_type = type(model_block[0])  # type: ignore
+    # When using the HF LM models,
+    # the weights of the self.lm_head and self.transformer.wte are tied.
+    # This tying occurs inside the `self.post_init()` function.
+    # This is a hurdle for FSDP because they need to be in the same FSDP block
+    # These lines ensures that both modules stay together in the top-most block when
+    # the model has this tying enabled (almost all do; this property defaults to True)
+    if model.config.tie_word_embeddings:
+        causal_base_model._fsdp_wrap = False  # type: ignore
+        tied_embeddings._fsdp_wrap = False  # type: ignore
+        lm_head._fsdp_wrap = False  # type: ignore
+
+    # FSDP Wrap and Activation Checkpoint every model block
+    model.fsdp_wrap_fn = lambda module: isinstance(module, block_type)
+    model.activation_checkpointing_fn = lambda module: isinstance(
+        module, block_type)
+
+
+def prepare_hf_enc_dec_model_for_fsdp(model: PreTrainedModel) -> None:
+    """Wrap an encoder/decoder HF model
+
+    This works for T5, BART, Pegasus, PegasusX, but not all enc/dec (ProphetNet)
+    You have model.shared, model.encoder, model.decoder and model.lm_head,
+    where model.shared are the embeddings which are tied to model.lm_head, and
+    model.shared == model.encoder.embed_tokens and
+    model.shared == model.decoder.embed_tokens
+    """
+    tied_embeddings = model.get_input_embeddings()
+    encoder = model.get_encoder()
+    decoder = model.get_decoder()
+    lm_head = hf_get_lm_head(model)
+    # some encoder/decoders have different layers for encoder vs decoder
+    encoder_block = hf_get_hidden_layers(encoder)
+    decoder_block = hf_get_hidden_layers(decoder)
+
+    modules = {
+        "encoder": encoder,
+        "decoder": decoder,
+        "encoder_block": encoder_block,
+        "decoder_block": decoder_block,
+        "lm_head": lm_head,
+        "tied_embeddings": tied_embeddings
+    }
+
+    for mod_name, module in modules.items():
+        if module is None:
+            raise ValueError(
+                f'Unable to FSDP-wrap this model! `{mod_name}` does not ' +
+                'follow common layer/weight naming conventions.')
+    decoder_block_type = type(decoder_block[0])  # type: ignore
+    encoder_block_type = type(encoder_block[0])  # type: ignore
+
+    if model.config.tie_word_embeddings:
+        # it is possible to train an enc/dec without tied embeddings, hence the check
+        tied_embeddings._fsdp_wrap = False  # type: ignore
+        encoder._fsdp_wrap = False  # type: ignore
+        decoder._fsdp_wrap = False  # type: ignore
+        lm_head._fsdp_wrap = False  # type: ignore
```

Comment:
Do we need to do anything special with `model.{encoder,decoder}.embed_tokens` since they are tied to `model.shared` and `model.lm_head`?

Reply:
I don't think so. `model.{encoder,decoder}.embed_tokens` are both references to `model.shared`; that is, `model.encoder.embed_tokens == model.shared`, and changing a property of the referent should change the 2 references

But, we could check this just to be safe.

That said, in the next iteration the plan is to do `isinstance(module, (encoder_bloc_type, decoder_block_type))` as the only check, which is what the [`transformer_auto_wrap_policy`](https://github.com/pytorch/pytorch/blob/c0e15fe3f4d5622c7a687cf0d503902be9296596/torch/distributed/fsdp/wrap.py#L81-L86) does